### ISSUE TITLE
[REQ] Java Spring generic ApiResponse

### ISSUE THREAD

**Issue by Quillraven** [OpenAPITools/openapi-generator#16428]
### Is your feature request related to a problem? Please describe.

I couldn't find a solution to define a generic response for open-api Spring generator or in general for the open-api generator.

## Describe the solution you'd like

In my yaml file I want to define something like this:
```
ApiResponse:
  type: object
  properties:
    result:
      type: object
    errorCode:
      type: integer
    errors:
      type: array
      items:
        type: string
```

and have specific endpoints return a specific implementation of the generic response like:

```
paths:
  /users:
    get:
      responses:
        content:
          'application/json':
            schema:
              $ref: '#/components/schemas/ApiResponse'
              result:
                type: User <-- somehow define the specific object type of result. Ideally a ref to another schema

  /books:
    get:
      responses:
        content:
          'application/json':
            schema:
              $ref: '#/components/schemas/ApiResponse'
              result:
                type: Book <-- somehow define the specific object type of result. Ideally a ref to another schema
```

This should then create a Java class for the ApiResponse like:
```
class ApiResponse<T> {

  private T result;
  private int errorCode;
  private List<String> errors;

}
```

and for the endpoints:

```
ApiResponse<User> getUsers() ...

ApiResponse<Book> getBooks() ...
```

## Describe alternatives you've considered

I thought of somehow using mustache templates and overriding specific schemas with the template but I don't know how that would work. Maybe there is already an easier build in solution?
Otherwise, please guide me to a useful documentation on how to use templates for this use-case.

## Additional context

...

**Comment by ranger-ross** [comment]
So as I understand it, the generic data structures support is part of [JSON Schema](https://json-schema.org/blog/posts/dynamicref-and-generics#using-dynamic-references-to-support-generic-types) which the OAS 3.1 is built upon. The only difference in 3.1.1 is the explicitly calls out the use of using `dynamicRef` and `dynamicAnchor` for generic data structures as an example. But technically the 3.1.0 spec already supports this according to the OAS maintainers (see [this](https://github.com/OAI/OpenAPI-Specification/pull/4082#issuecomment-2419911099) comment)

**Comment by ranger-ross** [comment]
Regardless, the 3.1.1 spec looks like it is going to land in the near future :smiley:

**Comment by LiyakhathSyed** [comment]
now that 3.1.1 is released. is progress being made here

**Comment by bsayli** [comment]
I ran into the same need—preserving **generic envelopes** like `ApiResponse<T>` (and even nested forms such as `ApiResponse<Page<T>>`) in generated Java clients—without waiting for OAS 3.1.1 features.

---

### 🔊 What worked for us (today, on OAS 3.1.0)

**Server side (OpenAPI producer):**

* Generate a composed schema and mark it with vendor extensions that tell the generator *what `T` is* and, when applicable, *what the container is*:

  ```yaml
  components:
    schemas:
      ApiResponseUser:
        allOf:
          - $ref: '#/components/schemas/ApiResponse'
          - type: object
            properties:
              data:
                $ref: '#/components/schemas/User'
        x-api-wrapper: true
        x-api-wrapper-datatype: User
        # optional for nested: x-data-container: Page / x-data-item: User
  ```

  We auto-add these via Springdoc customizers that introspect controller return types, so no hand-written subclasses are needed.

**Client side (OpenAPI Generator overlay):**

* Tiny Mustache override in `model.mustache`:

  ```mustache
  {{#vendorExtensions.x-api-wrapper}}
    {{>api_wrapper}}
  {{/vendorExtensions.x-api-wrapper}}
  {{^vendorExtensions.x-api-wrapper}}
    {{#isEnum}}{{>modelEnum}}{{/isEnum}}
    {{^isEnum}}{{>pojo}}{{/isEnum}}
  {{/vendorExtensions.x-api-wrapper}}
  ```
* And the partial `api_wrapper.mustache` renders a **thin shell** that keeps generics intact:

  ```mustache
  public class {{classname}}
      extends com.example.client.ServiceClientResponse<
        {{#vendorExtensions.x-data-container}}
          {{vendorExtensions.x-data-container}}<{{vendorExtensions.x-data-item}}>
        {{/vendorExtensions.x-data-container}}
        {{^vendorExtensions.x-data-container}}
          {{vendorExtensions.x-api-wrapper-datatype}}
        {{/vendorExtensions.x-data-container}}
      > {}
  ```

This yields:

```java
ApiResponseUser extends ServiceClientResponse<User>
ApiResponsePageUser extends ServiceClientResponse<Page<User>>
```

…so you get a single, reusable `ServiceClientResponse<T>` and strongly-typed payloads with no boilerplate duplication.

---

### 🔹 Why this helps

* **Type-safe generics** in generated clients (including nested `Page<T>`).
* **DRY envelope**: `{data, meta}` lives once in a generic base.
* **Minimal overlay**: one conditional + one tiny partial; default templates stay intact for everything else.

---

### 🔹 Runnable reference & write-up
Repo (runnable):
🔗 [blueprint-platform/openapi-generics](https://github.com/blueprint-platform/openapi-generics)
Article (why + how):
📖 [We Made OpenAPI Generator Think in Generics](https://medium.com/@baris.sayli/type-safe-generic-api-responses-with-spring-boot-3-4-openapi-generator-and-custom-templates-ccd93405fb04)

Happy to contribute a doc page or a small opt-in flag (e.g., `--additional-properties=useGenericWrappers=true`) if the project is open to documenting this pattern.

**Comment by mozinrat** [comment]
Yes now since we have spec version 3.2 released, what is the next steps by community. The above solution by @bsayli is great but somehow it relies on augmenting spec manually for client side usage. Also We Still generate dummy class like ServiceClientResponseUser. 

There is no solution on server side generation though which allow to generate reliably server stubs for spring for api first development style to support similar use case

**Comment by bsayli** [comment]
@mozinrat hello, replying to your earlier comment to clarify the core point:

The key issue is that OAS 3.1.1 / 3.2 only *documents* how generics can be expressed using JSON Schema (`$dynamicRef`),
but the Java / Spring / OpenAPI Generator toolchain still does not consume or emit those constructs end-to-end.

Until:
- Springdoc can emit `$dynamicRef` from `ResponseEntity<ApiResponse<T>>`
- OpenAPI Generator can map `$dynamicRef` to Java generics

any working solution must explicitly bridge this semantic gap.

The vendor-extension approach is not a workaround for the specification,
but a compatibility layer between:
- Java’s type system
- today’s OpenAPI tooling

Once the ecosystem supports `$dynamicRef` end-to-end, this layer can be removed.
Until then, this is the only production-grade, type-safe solution we’ve found.

**Comment by bsayli** [comment]
**Update (Apr 2026):** The approach described above is now available as a reusable Maven dependency — no template copying needed.

Both sides are covered:
- **Server**: `openapi-generics-server-starter` — auto-emits the vendor extensions via Springdoc customizers
- **Client**: `openapi-generics-java-codegen-parent` — handles the Mustache overlay as a build dependency

🔗 https://github.com/blueprint-platform/openapi-generics
